PR: Add new combobox widgets (API)

ccordoba12 opened the pull request:
## Description of Changes

- This allows us to improve the style of all comboboxes used in the application.
- UI improvements:
  - Add more padding to combobox items (it was almost non-existent).
  - Change hover color to the one used for any other widget in Spyder.
  - Remove visual glitch to the left of items for some comboboxes.
  - Fix misalignment of combobox popup on Mac.
  - Add padding around the combobox.
  - Paint combobox separators according to the interface theme.
  - Improve style of the class/function dropdown a bit by adding margins around it and between its comboboxes.
  - Make the QFontComboBox'es used in Preferences to follow the style for other comboboxes as well.

- UX improvements: 
   -  Disable context menu of editable comboboxes because it's not localized and it doesn't use our API.

- Other changes:
  - Replace all uses of raw Qt comboboxes with the new ones in this PR.

## Visual changes

**Before**

![imagen](https://github.com/spyder-ide/spyder/assets/365293/e83d71bd-d972-4e1e-ad39-0ae037a9a2ea)

![imagen](https://github.com/spyder-ide/spyder/assets/365293/3433c270-b14e-4ebc-83ab-c39c58ef9871)

![imagen](https://github.com/spyder-ide/spyder/assets/365293/e5f05d5b-8318-42d1-aee2-e79877fc02e5)

![imagen](https://github.com/spyder-ide/spyder/assets/365293/62f8211a-3d6a-404d-8063-227ec45e4cf4)

----

**After**

![imagen](https://github.com/spyder-ide/spyder/assets/365293/9e9828d5-9b7d-4220-afaa-03d7619a6d24)

![imagen](https://github.com/spyder-ide/spyder/assets/365293/0fc80468-6e1d-45a4-8598-e7444deee35f)

![imagen](https://github.com/spyder-ide/spyder/assets/365293/f2d52bfe-d86a-4c47-b6cb-d8e6f04c148f)

![imagen](https://github.com/spyder-ide/spyder/assets/365293/0db93ce3-4119-49d6-9a97-0df963cb6526)

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @ccordoba12 

<!--- Thanks for your help making Spyder better for everyone! --->

ccordoba12 commented:
@dalthviz, this is ready for review. Could you also check how things look on Windows and Mac? I wouldn't expect big surprises there, but you never know.

pep8speaks commented:
Hello @ccordoba12! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:































* In the file [`spyder/utils/stylesheet.py`](https://github.com/spyder-ide/spyder/blob/5f389248758e2a0dd3871f6e92d9ec9b93fb9f37/spyder/utils/stylesheet.py):

> [Line 39:19](https://github.com/spyder-ide/spyder/blob/5f389248758e2a0dd3871f6e92d9ec9b93fb9f37/spyder/utils/stylesheet.py#L39): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 42:23](https://github.com/spyder-ide/spyder/blob/5f389248758e2a0dd3871f6e92d9ec9b93fb9f37/spyder/utils/stylesheet.py#L42): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 43:20](https://github.com/spyder-ide/spyder/blob/5f389248758e2a0dd3871f6e92d9ec9b93fb9f37/spyder/utils/stylesheet.py#L43): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 46:27](https://github.com/spyder-ide/spyder/blob/5f389248758e2a0dd3871f6e92d9ec9b93fb9f37/spyder/utils/stylesheet.py#L46): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

dalthviz commented:
Did an initial check on Windows and seems like things are looking good except for the following (although not totally sure if is caused due to this PR or something that was happening before):

* In the working directory combobox select a path so the validation is done (a check appears). Seems like the line edit overflows vertically the combobox (bottom blue border of the combobox is not visible):

![image](https://github.com/spyder-ide/spyder/assets/16781833/a536b2c2-7676-4d33-b76f-017314f18710)

* Seems like there are some extra division line in the Find in Files combox under the `Clear this list` action (although I think that happens on Linux too checking the OP screenshots):

![image](https://github.com/spyder-ide/spyder/assets/16781833/202b95f6-9138-42e8-885b-9abe5004d002)

On macOS there is a more or less noticeable misalignment for the comboboxes- The Find in Files `Search in:` is the one that show a more prominent misalignment:

![image](https://github.com/spyder-ide/spyder/assets/16781833/43149791-c293-441b-b162-939499b992c5)

![image](https://github.com/spyder-ide/spyder/assets/16781833/342ef056-54a7-47f2-bd09-497ae75828a2)

ccordoba12 commented:
Thanks @dalthviz for the review! About your comments:

> In the working directory combobox select a path so the validation is done (a check appears). Seems like the line edit overflows vertically the combobox (bottom blue border of the combobox is not visible)

Fixed.

> Seems like there are some extra division line in the Find in Files combox under the Clear this list action (although I think that happens on Linux too checking the OP screenshots)

I plan to fix that after this PR is merged. Actually, my initial motivation for this work was to fix that problem.

> On macOS there is a more or less noticeable misalignment for the comboboxes

Fixed.

I also make the two QFontCombox'es we are using in Preferences to follow the style of the other comboboxes. See the updated OP for the screenshots.

dalthviz commented:
Checked again with the latest changes on Windows and now seems like some comboboxes that have an initial value appears with a white background?

This is what I'm seeing for the Help pane `Source` combobox:

![image](https://github.com/spyder-ide/spyder/assets/16781833/f42ae81f-6d19-42ec-af60-244ac4acdd4d)

And it also happens with the Find pane `Search in:` combobox. Seems like after a first interaction the background gets the correct background?:

![combobox_find](https://github.com/spyder-ide/spyder/assets/16781833/01b45dea-7d83-4d1e-8373-3e837655214e)

Also, checked the working directory combobox but it still shows the same behavior :thinking:

![combobox_cwd](https://github.com/spyder-ide/spyder/assets/16781833/5ce99c87-958d-4f0b-8453-2363cf3835d3)

Also, regarding the comboboxes to select fonts seems like now the items doesn't show a preview of the font? Is not that kind of a regression or maybe for that kind of combobox a new preview widget is meant to be done?

![combobox_font](https://github.com/spyder-ide/spyder/assets/16781833/ebe7466a-3bd3-4992-bc7e-17a09dc37462)

dalthviz commented:
Did a quick check on macOS and seems like there things regarding alignment are correct now :+1:

![image](https://github.com/spyder-ide/spyder/assets/16781833/4c3ceffa-1a88-46d1-b864-7625cfea37d5)

ccordoba12 commented:
> Checked again with the latest changes on Windows and now seems like some comboboxes that have an initial value appears with a white background?

I was able to reproduce this and it should be fixed in my last commits.

> Also, checked the working directory combobox but it still shows the same behavior 🤔

Really fixed now. I was not seeing this problem on Linux, so I thought some of my previous changes should have fixed it.

> Also, regarding the comboboxes to select fonts seems like now the items doesn't show a preview of the font? Is not that kind of a regression or maybe for that kind of combobox a new preview widget is meant to be done?

I understand your point of view but I was not able to find a way to set a specific font for combobox items. And I think the current style doesn't look good:

![image](https://github.com/spyder-ide/spyder/assets/365293/bfe29e29-1334-4a94-9444-96d8bfed253b)

Furthermore, I also checked what RStudio does and it simply shows the font names, as we are doing now.

dalthviz commented:
Checked again and seems like things are working now :+1: Regarding the fonts comboboxes, I would suggest to make the `Monospace` font setting value changes affect the `Preview` that is already in that preference page. Going even further probably the font size preference should have a similar behavior (alter the `Preview` text when its value changes). A similar interaction can be seen in the [Qt `QFontDialog`](https://doc.qt.io/qt-6/qfontdialog.html) where a `Sample` widget shows how the font looks:

![combobox_font_2](https://github.com/spyder-ide/spyder/assets/16781833/54be5bb4-4b3c-4cca-a6ce-c10b74909ce3)

Following that idea I would also suggest to create a second preview widget for the `Interface` font. This could be useful since without any kind of preview now you most probably will end up having to restart Spyder multiple times to see how the different fonts look.

ccordoba12 commented:
> Checked again and seems like things are working now 👍

Great!

> Regarding the fonts comboboxes, I would suggest to make the Monospace font setting value changes affect the Preview that is already in that preference page. Going even further probably the font size preference should have a similar behavior (alter the Preview text when its value changes).

Great suggestion @dalthviz! I haven't thought about doing something like that, but I agree it'd help users to understand how font changes would affect the interface.

So, I implemented that in my last commit.

> Following that idea I would also suggest to create a second preview widget for the Interface font. This could be useful since without any kind of preview now you most probably will end up having to restart Spyder multiple times to see how the different fonts look.

That's a good suggestion too, but I think out of the scope for this PR. I have some ideas on how to implement it but for that I'd require PR #21511 merged first. So, I think it's better to leave it for later.

dalthviz commented on spyder/api/widgets/comboboxes.py:
```diff
@@ -0,0 +1,250 @@
+        self._css = self._generate_stylesheet()
+        self.setStyleSheet(self._css.toString())
+
+    def contextMenuEvent(self, event):
+        # Prevent showing context menu for editable comboboxes because it's
+        # added automatically by Qt. That means that the menu is not built
+        # using our API and it's not localized.
+        pass
```
Should then a menu be manually created instead of just passing? WIthout an implementation then the only way you have to copy/paste/cut text is by using shortcuts, no?

ccordoba12 replied:
I talked about this to @conradolandia in one of our UX meetings and we agreed that:

* It's a minor functionality that can also be done using the mouse or keyboard shortcuts (as you said).
* It's not worth it to invest time to implement a new, localized menu that uses our API because we have a ton of other things to do for Spyder 6.

dalthviz reviewed (commented):
Thanks @ccordoba12 ! Just left a comment regarding completly removing the context menu from comboboxes other than that this LGTM :+1:

dalthviz reviewed (approved):
Thanks @ccordoba12 !

dalthviz commented:
Just in case, is there anything else you want to add here @ccordoba12 ? I just noticed that this is marked for beta1 🤔

ccordoba12 commented:
Nop. I thought it was going to take longer to get this PR approved, so I didn't want to delay alpha3 due to that. But since you think it's ready, let's include in that release.
